Decorators

WareJosephB opened the pull request:
- [x] Linting
- [x] Tests
- [x] New tests for new functionality
- [x] Useful change

Commit:
Merge remote-tracking branch 'worigin/main' into main

# Conflicts:
#	docs/how-to/construct-a-control-object.rst
#	src/python_bluesky_taskgraph/core/decision_engine.py

Commit:
Adjust PyPi API secret name

First release published to PyPi

Commit:
Add first decorators, allow single tasks to be added to Graphs

Commit:
Add first decorators, allow single tasks to be added to Graphs

Commit:
Merge branch 'Decorators' of https://github.com/DiamondLightSource/python-bluesky-taskgraph into Decorators

Commit:
Merge branch 'Decorators' of https://github.com/DiamondLightSource/python-bluesky-taskgraph into Decorators

Commit:
Merge branch 'Decorators' of https://github.com/DiamondLightSource/python-bluesky-taskgraph into Decorators

callumforrester commented on src/python_bluesky_taskgraph/core/decision_engine.py:
```diff
@@ -60,6 +57,7 @@ def __init__(
         self._should_stop_at_end_of_next_run: bool = False
         self._error_tasks: Dict[str, int] = {}
         self._recovered_tasks: Set[str] = set()
+        self._logger = logging.getLogger(self.__class__.__name__)
```
Can have loggers at the module level rather than the class level for less boilerplate, so at the top of the file:

```python
LOGGER = logging.getLogger(__name__)
```

WareJosephB replied:
Depends on whether there'd ever be more than one DecisionEngine, [of different types] that we might want to distinguish logging for. I think with the same argument as having a single RunEngine, that I oppose that so I'm happy to make this change: and it simplifies the logging exception handling.

callumforrester commented on src/python_bluesky_taskgraph/core/task.py:
```diff
@@ -48,15 +51,14 @@ class BlueskyTask(Generic[InputType]):
         self._name: str = name
-        # TODO: Do we want logging at the Task, DecisionEngine or ControlObject level?
-        self._logger = logging.getLogger("BlueskyTask")
+        self._logger = logging.getLogger(self.name)
```
Ditto for logger

WareJosephB replied:
Tasks definitely should log on the level of individual instances of the task. Have changed to extend the module level logger, so logs can be bundled or split apart appropriately.

callumforrester commented on src/python_bluesky_taskgraph/core/task_graph.py:
```diff
@@ -1,13 +1,29 @@
+
+
+@dataclass
+class TaskTuple:
```
Might think about renaming this, only because `Tuple` is an overloaded word (`NamedTuple` et al.) and I got confused

```suggestion
class PreparedTask:
```

callumforrester commented on src/python_bluesky_taskgraph/core/task_graph.py:
```diff
@@ -1,13 +1,29 @@
+@dataclass
+class TaskTuple:
+    task: BlueskyTask
+    inputs: List[str]
```
Does this relate to `TypedInput` below?

WareJosephB replied:
No relation to TypedInput, TypedInput is a utility type to facilitate Tasks with a single argument (and might be left on the cutting room floor before too long). e.g. 

PlantRequiringTask(BlueskyTask[TypedInput[Plant]]): as "short"cut for 

PlantRequiringTask(BlueskyTask['PlantInput']): 
  @dataclass 
  PlantInput: 
    plant: Plan


TaskTuple/PreparedTask just triplet of task, input names, output names. Always List[str].

callumforrester commented on src/python_bluesky_taskgraph/core/types.py:
```diff
@@ -1,5 +1,5 @@
 from dataclasses import dataclass, field
```
I recommend renaming this file because `types` is a module in the standard library.
If it's private (i.e. only used within `core`) you could call it `_types.py`.

WareJosephB replied:
Done (-> type_hints as some needed external to core)

callumforrester commented on src/python_bluesky_taskgraph/tasks/functional_tasks.py:
```diff
@@ -63,7 +63,7 @@ class PlanTask(BlueskyTask["PlanTask.PlanInputs"]):
     @dataclass
     class PlanInputs(Input):
-        args: List[Any]
+        args: List[Any] = field(default_factory=list)
```
Ditto, does this need to be updated wrt `TypedInput`?

WareJosephB replied:
See above

Most plans aren't single argument. PlanTask going to face big refactor when I approach it likely.

callumforrester commented on src/python_bluesky_taskgraph/core/task.py:
```diff
@@ -147,3 +150,29 @@ def _overwrite_results(self, results: List[Any] = None) -> None:
+def run_stage_decorator(
+    func: Callable[[InputType], TaskOutput]
+) -> Callable[[InputType], TaskOutput]:
+    def decorated_func(args: InputType):
```
Should this be?
```suggestion
    def decorated_func(args: InputType) -> TaskOutput:
```

WareJosephB replied:
Done

callumforrester commented on src/python_bluesky_taskgraph/core/task.py:
```diff
@@ -147,3 +150,29 @@ def _overwrite_results(self, results: List[Any] = None) -> None:
+    func: Callable[[InputType], TaskOutput]
+) -> Callable[[InputType], TaskOutput]:
+    def decorated_func(args: InputType):
+        devices = {device for device in astuple(args) if isinstance(device, Device)}
```
We don't want to tie this to the concrete `Device` as this will break when we use a non-ophyd library (such as ophyd v2). 
Suggest using the `bluesky.protocols.Stageable` interface instead.

```suggestion
        devices = {device for device in astuple(args) if isinstance(device, Stageable)}
```

WareJosephB replied:
Done

callumforrester replied:
Idea for future refactor would be to remove ophyd references entirely from `core` and just use the interfaces but that's out-of-scope.

callumforrester commented on src/python_bluesky_taskgraph/core/task.py:
```diff
@@ -147,3 +150,29 @@ def _overwrite_results(self, results: List[Any] = None) -> None:
+    def wrapper_stage_decorator(*args, **kwargs) -> BlueskyTask[InputType]:
+        task: BlueskyTask = func(*args, **kwargs)
+        # Prevents MyPy complaints about assigning to a function
+        task.__setattr__("_run_task", run_stage_decorator(task._run_task))
```
Not too bothered about this, but alternative is to make a wrapper task

WareJosephB replied:
Harder to test, as the passing of the status between a wrapping mock and the underlying task leads to problems passing completion around. I do think it's a neater way of doing it though, implementation similar to ConditionalTask.

callumforrester reviewed (commented):
Might have got confused in a few places, sorry if so!

Commit:
Make adjustments from merge review

- Rename types -> type_hints to prevent namespace collisions
- Refactor logging to more logical and hierarchical form
- Extend Staging decoator to Stageable protocol
- Rename TaskTuple -> PreparedTask

Commit:
Make adjustments from merge review

- Rename types -> type_hints to prevent namespace collisions
- Refactor logging to more logical and hierarchical form
- Extend Staging decoator to Stageable protocol
- Rename TaskTuple -> PreparedTask

Commit:
Merge remote-tracking branch 'origin/Decorators' into Decorators

Commit:
Merge remote-tracking branch 'origin/Decorators' into Decorators

Commit:
Merge remote-tracking branch 'origin/Decorators' into Decorators
